proposals/p3848.md

```diff
@@ -0,0 +1,576 @@
+
+fn Foo4() {
+  fn Bar4 {
+    Print($0);  // ✅ Valid: Foo4 has explicit parameters
+  }
+}
+
+fn Foo5 {
+  fn Bar5() {}  // ✅ Valid: Bar5 has explicit parameters
+}
+
+fn Foo6() {
+  my_list.Sort(
+    fn => $0 < $1  // ✅ Valid: Foo6 has explicit parameters
+  );
+}
+```
+
+**Alternative Considered**: In addition to the proposed restrictions, an
+additional restriction was considered. That being, visibility of functions with
+positional parameters could be restricted to only non-public interfaces. **This
+alternative will be put forth as a leads question before a decision is made.**
+
+## Function Captures
+
+**Proposal**: Function captures in Carbon mirror the non-init captures of C++.
+Bindings defined in a scope visible to a function can be captured in the
+function instance, be it a function declaration or a lambda. The lifetime of a
+capture is the lifetime of the function in which it exists. For example...
+
+```
+fn Foo() {
+  let handle: Handle = Handle.Get();
+  var thread: Thread = Thread.Make(fn [copy handle] { handle.Process(); });
+  thread.Join();
+}
+```
+
+```
+fn Foo() {
+  let handle: Handle = Handle.Get();
+  fn MyThread[copy]() { handle.Process(); }
+  var thread: Thread = Thread.Make(MyThread);
+  thread.Join();
+}
+```
+
+### Default Capture Mode
+
+**Proposal**: By default, there is no capturing in functions. The lack of any
+square brackets is the same as an empty pair of square brackets. Users can opt
+into capturing behavior. This is done either by way of individual explicit
+captures, or more succinctly by way of a default capture mode. The default
+capture mode mirrors the syntax `[=]` and `[&]` capture modes from C++ (written
+in Carbon as `[copy]` and `[ref]`) by being the first thing to appear in the
+square brackets.
+
+```
+fn Foo() {
+  let handle: Handle = Handle.Get();
+  fn MyThread[copy]() {
+    handle.Process();  // `handle` is copy captured due to the default capture
+                       // mode specifier of `copy`
+  }
+  var thread: Thread = Thread.Make(MyThread);
+  thread.Join();
+}
+```
+
+**Alternative Considered**: Previously, it was proposed that the default capture
+mode would come after all the explicit captures. In addition, it was proposed
+that the lack of any square brackets opted the function into capturing by
+default with an implicitly defined default capture mode. These behaviors do not
+mirror lambdas in C++ and so were decided against. Primarily, it was recognized
+that it's valuable to be able to intermix explicit captures with deduced
+parameters and fields in any order that makes the most sense for the context.
+Without a clear justification for a rule that says they can't intermix, the more
+flexible behavior was favored.
+
+### Capture Modes
+
+**Proposal**: `copy` and `ref` are the two primary forms of function captures.
+They each have companion `const` versions. They are syntactically distinct from
+`let` and `var`. `let` and `var` can appear in binding patterns, such as
```

Comment:
It feels tautological to say that `copy` and `ref` are syntactically distinct from `let` and `var`: they're different tokens, so of course they're syntactically distinct. I'd suggest instead saying that function captures are syntactically distinct from function fields. And probably that should go in the Function Captures section, because it seems more of a fact about function captures as a whole, not so much about the mode specifier tokens in particular.